src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,95 +248,108 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(ReadOnlySpan<byte> buffer)
         {
-            Debug.Assert(buffer != null);
-            Debug.Assert(buffer.Length > 0);
+            ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
-            fixed (byte* b = buffer)
+            // Parse each event from the buffer and notify appropriate delegates
+            for (;;)
             {
-                Interop.Kernel32.FILE_NOTIFY_INFORMATION* info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
+                // Validate the data we received in case it's corrupted. This can happen
+                // if we are watching files over the network with a poor connection.
 
-                ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
+                // Verify the info object is within the expected bounds
+                if (sizeof(FILE_NOTIFY_INFORMATION) > (uint)buffer.Length)
+                {
+                    break;
```

Comment:
Per your comment in the other issue, shouldn't we be asserting here and break statements below?  There's no documented behavior of ReadDirectoryChanges that it actually will return truncated results.  We've just seen it in practice.

Reply:
Added comments and asserts.